url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/term/Function.java

```diff
@@ -0,0 +1,689 @@
+import com.fuzzylite.variable.InputVariable;
+import com.fuzzylite.variable.OutputVariable;
+import java.lang.reflect.Method;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.StringTokenizer;
+import java.util.logging.Logger;
+
+/**
+ *
+ * @author jcrada
+ */
+public class Function extends Term {
+
+    public static abstract class Element {
+
+        public String name;
+        public Method method;
+
+        public int getArity() {
+            return method.getParameterTypes().length;
+        }
+    }
+
+    public static class Operator extends Element {
+
+        public int precedence, associativity;
+
+        public Operator() {
+            this("", null, 0);
+        }
+
+        public Operator(String name, Method operator, int precedence) {
+            this(name, operator, precedence, -1);
+        }
+
+        public Operator(String name, Method operator, int precedence, int associativity) {
+            this.name = name;
+            this.method = operator;
+            this.precedence = precedence;
+            this.associativity = associativity;
+        }
+    }
+
+    public static class BuiltInFunction extends Element {
+
+        public int associativity;
+
+        public BuiltInFunction() {
+            this("", null);
+        }
+
+        public BuiltInFunction(Method function) {
+            this(function.getName(), function);
+        }
+
+        public BuiltInFunction(String name, Method function) {
+            this(name, function, -1);
+        }
+
+        public BuiltInFunction(String name, Method function, int associativity) {
+            this.name = name;
+            this.method = function;
+            this.associativity = associativity;
+        }
+    }
+
+    public static class Node {
+
+        public Operator operator = null;
+        public BuiltInFunction function = null;
+        public String variable = "";
+        public double value = Double.NaN;
+        public Node left = null, right = null;
+
+        public Node(Operator operator) {
+            this.operator = operator;
+        }
+
+        public Node(BuiltInFunction function) {
+            this.function = function;
+        }
+
+        public Node(String variable) {
+            this.variable = variable;
+        }
+
+        public Node(double value) {
+            this.value = value;
+        }
+
+        public double evaluate(Map<String, Double> localVariables) {
+            double result = Double.NaN;
+            if (operator != null || function != null) {
+                Element element = function;
+                if (operator != null) {
+                    element = operator;
+                }
+
+                try {
+                    switch (element.getArity()) {
+                        case 0:
+                            result = (double) element.method.invoke(null);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)